src/projections.jl

```diff
@@ -152,6 +158,116 @@ function vec_symm(X)
     return X[LinearAlgebra.triu(trues(size(X)))]
 end
 
+"""
+    projection_on_set(::DefaultDistance, v::AbstractVector{T}, ::MOI.ExponentialCone) where {T}
+
+projection of vector `v` on closure of the exponential cone
+i.e. `cl(Kexp) = {(x,y,z) | y e^(x/y) <= z, y>0 } U {(x,y,z)| x <= 0, y = 0, z >= 0}`.
+
+References:
+* [Proximal Algorithms, 6.3.4](https://web.stanford.edu/~boyd/papers/pdf/prox_algs.pdf)
+by Neal Parikh and Stephen Boyd.
+* [Projection, presolve in MOSEK: exponential, and power cones]
+(https://docs.mosek.com/slides/2018/ismp2018/ismp-friberg.pdf) by Henrik Friberg
+"""
+function projection_on_set(::DefaultDistance, v::AbstractVector{T}, s::MOI.ExponentialCone) where {T}
+    _check_dimension(v, s)
+
+    if _in_exp_cone(v; dual=false)
+        return v
+    elseif _in_exp_cone(-v; dual=true)
+        # if in polar cone Ko = -K*
+        return zeros(3)
+    elseif v[1] <= 0 && v[2] <= 0
+        return [v[1]; 0.0; max(v[3],0)]
+    else
+        return _exp_cone_proj_case_4(v)
+    end
+end
+
+function _in_exp_cone(v::AbstractVector{T}; dual=false) where {T}
+    # See pg. 184 https://web.stanford.edu/~boyd/papers/pdf/prox_algs.pdf
+    # TODO: Tol for == 0 to avoid denom blowing up? see case 4 in deriv
+    if dual
+        return (
+            (v[1] == 0 && v[2] >= 0 && v[3] >= 0) ||
+            (v[1] < 0 && v[1]*exp(v[2]/v[1]) + ℯ*v[3] >= EXP_CONE_THRESH)
+        )
+    else
+        return (
+            (v[1] <= 0 && v[2] == 0 && v[3] >= 0) ||
+            (v[2] > 0 && v[2] * exp(v[1] / v[2]) - v[3] <= EXP_CONE_THRESH)
+        )
+    end
+end
+
+function _exp_cone_proj_case_4(v::AbstractVector{T}) where {T}
+    # https://docs.mosek.com/slides/2018/ismp2018/ismp-friberg.pdf
+    # Thm: h(x) is smooth, strictly increasing, and changes sign on domain
+    r, s, t = v[1], v[2], v[3]
+    h(x,p) = (((x-1)*r + s) * exp(x) - (r - x*s)*exp(-x))/(x^2 - x + 1) - t
+
+    # Note: won't both be Inf by case 3 of projection
+    lb = r > 0 ? 1 - s/r : -Inf
+    ub = s > 0 ? r/s : Inf
+
+    # Deal with ±Inf bounds
+    if isinf(lb)
+        lb = min(ub-0.125, -0.125)
+        for _ in 1:10
+            h(lb, nothing) < 0 && break
+            ub = lb
+            lb *= 2
+        end
+    elseif isinf(ub)
+        ub = max(lb+0.125, 0.125)
+        for _ in 1:10
+            h(ub, nothing) > 0 && break
+            lb = ub
+            ub *= 2
+        end
+    end
+
+    if !(h(lb, nothing) < 0 && h(ub, nothing) > 0)
+        error("Failure to find bracketing interval for exp cone")
+    end
+
+    # Solve with Bisection
+    prob = NonlinearSolve.NonlinearProblem(h, (lb, ub))
+    sol = NonlinearSolve.solve(prob, NonlinearSolve.Bisection())
```

Comment:
If you're just doing bisection, it's probably easier to just write than to include a heavy dependency.

Reply:
Just added in the latest edit